### PR TITLE
#10129: Move rotate_half, rotary_embedding/llama; split transformers

### PR DESCRIPTION
**Pull request by ayerofieiev-tt**
### Ticket
#10128
#10129 

### Problem description
Moving rotate_half and rotary_embedding

### What's changed
Moved rotate_half and rotary_embedding and rotary_embedding_llama
Splitted transformer.hpp

### Checklist
- [x] Post commit CI [passes](https://github.com/tenstorrent/tt-metal/actions/runs/10362825963)
- [x] t3k demo tests [passes](https://github.com/tenstorrent/tt-metal/actions/runs/10362352763)
- [x] t3k frequent tests [passes](https://github.com/tenstorrent/tt-metal/actions/runs/10362354051)

### CODE REVIEW

File: ttnn/CMakeLists.txt [hunk 1]
```diff
@@ -1,151 +1,221 @@
...
+set(NON_PYBIND_SRC)
+set(PYBIND_SRC)
+foreach(FILE ${TTNN_SRCS})
+    if(FILE MATCHES "_pybind.cpp$")
+        list(APPEND PYBIND_SRC ${FILE})
+    else()
+        list(APPEND NON_PYBIND_SRC ${FILE})
+    endif()
+endforeach()
+set(TTNN_SRCS ${NON_PYBIND_SRC})
```
Comment (ayerofieiev-tt) [inline ttnn/CMakeLists.txt:216]: @yan-zaretskiy fyi

**Review by TT-BrianLiu — APPROVED** [review]

**Review by arakhmati — APPROVED** [review]

File: ttnn/CMakeLists.txt [hunk 2]
```diff
@@ -1,151 +1,221 @@
...
+    ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/transformer/split_query_key_value_and_split_heads/split_query_key_value_and_split_heads_pybind.cpp
+
+    ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/operations/transformer/transformer_pybind.cpp
 )
 
+# Split src and python bindings
+# We only build python bindings optionally
+set(NON_PYBIND_SRC)
+set(PYBIND_SRC)
+foreach(FILE ${TTNN_SRCS})
```
Comment (yan-zaretskiy) [inline ttnn/CMakeLists.txt:209]: I would use a different variable name here from `TTNN_SRCS` set on line 216, just for clarity.

File: ttnn/cpp/ttnn/operations/transformer/split_query_key_value_and_split_heads/split_query_key_value_and_split_heads.cpp [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+#include "ttnn/deprecated/tt_dnn/op_library/nlp_tms/nlp_tms.hpp"
+
+#include "ttnn/operations/core/core.hpp"
+
+
+namespace ttnn::operations::transformer {
+
+namespace detail {
+inline std::tuple<Tensor, Tensor, Tensor> reshape_outputs_of_split_query_key_value_and_split_heads(
```
Comment (yan-zaretskiy) [inline ttnn/cpp/ttnn/operations/transformer/split_query_key_value_and_split_heads/split_query_key_value_and_split_heads.cpp:15]: Don't need `inline` here.

**Review by yan-zaretskiy — APPROVED** [review]

File: ttnn/CMakeLists.txt (hunk 2 above)
Comment (ayerofieiev-tt) [reply ttnn/CMakeLists.txt]: Done

File: ttnn/cpp/ttnn/operations/transformer/split_query_key_value_and_split_heads/split_query_key_value_and_split_heads.cpp (hunk 3 above)
Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/transformer/split_query_key_value_and_split_heads/split_query_key_value_and_split_heads.cpp]: fixed

**Review by skhorasganiTT — APPROVED** [review]

**Review by mtairum — APPROVED** [review]

**Review by cglagovichTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding/rotary_embedding.cpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+ttnn::Tensor RotaryEmbeddingOperation::operator()(
+    const Tensor& input_tensor,
+    const Tensor& cos_cache,
+    const Tensor& sin_cache,
+    const std::optional<uint32_t> token_index,
+    const std::optional<MemoryConfig> memory_config,
+    const std::optional<const DeviceComputeKernelConfig> compute_kernel_config) {
+
+    TT_FATAL(input_tensor.get_legacy_shape()[-1] % (TILE_WIDTH * 2) == 0, "Input X dim must be divisible into tiles");
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding/rotary_embedding.cpp:51]: Please provide better error message. Maybe include TILE_WIDTH.

File: ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding/rotary_embedding.cpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    uint32_t seq_len = input_tensor.get_legacy_shape()[-2];
+    uint32_t B = input_tensor.get_legacy_shape()[0];
+    uint32_t X = input_tensor.get_legacy_shape()[-1];
+    TT_FATAL(cos_cache.get_legacy_shape() == sin_cache.get_legacy_shape(), "Cos and Sin dims must match");
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding/rotary_embedding.cpp:55]: Not sure, up to you but we can add dimmensions to the explanation.

File: ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding/rotary_embedding.cpp [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    TT_FATAL(cos_cache.get_legacy_shape()[0] == 1 && cos_cache.get_legacy_shape()[1] == 1 && cos_cache.get_legacy_shape()[-1] == X, "Cos dims must match input dims");
+    if (token_index.has_value()) {
+        seq_len = input_tensor.get_legacy_shape()[0];
+        TT_FATAL(seq_len == 1);
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding/rotary_embedding.cpp:59]: I dont get why it should be one, and better to add error message.

File: ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding_llama/rotary_embedding_llama.cpp [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+#include "device/rotary_embedding_llama_device_operation.hpp"
+
+namespace ttnn::operations::experimental::transformer {
+
+Tensor RotaryEmbeddingLlamaOperation::operator()(
+    const Tensor &input_tensor,
+    const Tensor &cos,
+    const Tensor &sin,
+    const Tensor& trans_mat,
+    const std::optional<MemoryConfig> memory_config,
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding_llama/rotary_embedding_llama.cpp:16]: can we pass it by reference?

File: ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding_llama/rotary_embedding_llama.cpp [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    std::optional<const DeviceComputeKernelConfig> compute_kernel_config) {
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding_llama/rotary_embedding_llama.cpp:17]: same

File: ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding_llama/rotary_embedding_llama.cpp (hunk 7 above)
Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding_llama/rotary_embedding_llama.cpp:16]: thank you

File: ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding_llama/rotary_embedding_llama.cpp [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+#include "rotary_embedding_llama.hpp"
+
+#include "device/rotary_embedding_llama_device_operation.hpp"
+
+namespace ttnn::operations::experimental::transformer {
+
+Tensor RotaryEmbeddingLlamaOperation::operator()(
+    const Tensor &input_tensor,
+    const Tensor &cos,
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding_llama/rotary_embedding_llama.cpp:13]: I'd use cos_cache and sin_cache. using namespace std might be somewhere near  don't want to see weird errors in future :D

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding_llama/rotary_embedding_llama.cpp:13]: 100% actually replaced it in the other one, but forgot here. Thank you

File: ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding/rotary_embedding.cpp (hunk 6 above)
Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding/rotary_embedding.cpp:30]: I am not sure. This code as most of it is moved from tt_dnn, and I don't have full context

**Review by dmakoviichuk-tt — APPROVED** [review]

**Pull request merged**